### PR TITLE
[receiver/filelog] fix record counting with header

### LINKED ISSUE #35869: [receiver/filelog] record numbers are incorrect when `header` is configured
### Component(s)

receiver/filelog

## Description

Configuring the `header` option for the File Log receiver together with `include_file_record_number` causes the reported record numbers to be higher than they should be. This happens even when there are no header lines in the file.

## Steps to Reproduce (input file with header)

1. Prepare input file

   ```console
   cat > logs.txt <<EOF
   heredoc> #xyz
   heredoc> my-log
   heredoc> EOF
   ```
1. Run the collector with the below config `otelcol-contrib --config config.yaml`

## Expected Result

I would expect the "my-log" log to have `log.file.record_number` equal to `1`.

## Actual Result

The log has `log.file.record_number` equal to `3`.

The output from Debug exporter shows the following log:

```console
my-log log.file.name=logs.txt log.file.record_number=3 headerattr=xyz
```

## Steps to Reproduce (input file without header)

1. Prepare input file

   ```console
   cat > logs.txt <<EOF
   heredoc> my-log
   heredoc> EOF
   ```
1. Run the collector with the below config `otelcol-contrib --config config.yaml`

## Expected Result

I would expect the "my-log" log to have `log.file.record_number` equal to `1`.

## Actual Result

The log has `log.file.record_number` equal to `2`.

The output from Debug exporter shows the following log:

```console
my-log log.file.name=logs.txt log.file.record_number=2
```

### Collector version

v0.111.0

### Environment information

_No response_

### OpenTelemetry Collector configuration

```yaml
exporters:
  debug:
    verbosity: normal
receivers:
  filelog:
    header:
      metadata_operators:
        - type: regex_parser
          regex: ^#(?P<headerattr>\S+)$
      pattern: ^#
    include:
      - logs.txt
    include_file_record_number: true
    start_at: beginning
service:
  pipelines:
    logs:
      exporters:
        - debug
      receivers:
        - filelog
```

### Log output

```shell
2024-10-18T15:42:50.737+0200    info    service@v0.111.0/service.go:136 Setting up own telemetry...
2024-10-18T15:42:50.737+0200    info    telemetry/metrics.go:70 Serving metrics {"address": "localhost:8888", "metrics level": "Normal"}
2024-10-18T15:42:50.737+0200    info    builders/builders.go:26 Development component. May change in the future.        {"kind": "exporter", "data_type": "logs", "name": "debug"}
2024-10-18T15:42:50.739+0200    info    service@v0.111.0/service.go:208 Starting otelcol-contrib...     {"Version": "0.111.0", "NumCPU": 20}
2024-10-18T15:42:50.739+0200    info    extensions/extensions.go:39     Starting extensions...
2024-10-18T15:42:50.739+0200    info    adapter/receiver.go:47  Starting stanza receiver        {"kind": "receiver", "name": "filelog", "data_type": "logs"}
2024-10-18T15:42:50.740+0200    info    service@v0.111.0/service.go:234 Everything is ready. Begin running and processing data.
2024-10-18T15:42:50.940+0200    info    fileconsumer/file.go:256        Started watching file   {"kind": "receiver", "name": "filelog", "data_type": "logs", "component": "fileconsumer", "path": "logs.txt"}
2024-10-18T15:42:51.041+0200    info    LogsExporter    {"kind": "exporter", "data_type": "logs", "name": "debug", "resource logs": 1, "log records": 1}
2024-10-18T15:42:51.041+0200    info    my-log log.file.name=logs.txt log.file.record_number=3 headerattr=xyz
        {"kind": "exporter", "data_type": "logs", "name": "debug"}
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by andrzej-stencel**
#### Description

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35869 by refactoring of the `Reader::ReadToEnd` method.

This refactors the `Reader::ReadToEnd` method by separating reading the file's header from reading the file's contents.
This results in very similar code in `readHeader` and `readContents` methods, which was previously deduplicated at the cost of slightly higher complexity.
The bug could be fixed without separating header reading from contents reading, but I hope this separation will make it easier to implement content batching in the Reader (https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35455).
Content batching was my original motivation for these code changes.
I only discovered the problem with record counting when reading the code.

#### Link to tracking issue

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35869

#### Testing

In the first commit I have added tests that document the erroneous behavior. In the second commit I have fixed the bug and corrected the tests.

#### Documentation

Added changelog entry.

### CODE REVIEW

File: pkg/stanza/fileconsumer/internal/reader/reader.go [hunk 1]
```diff
@@ -106,9 +106,90 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
...
+	if err := r.headerReader.Stop(); err != nil {
+		r.set.Logger.Error("failed to stop header pipeline during finalization", zap.Error(err))
+	}
+	r.headerReader = nil
+	r.HeaderFinalized = true
+	r.initialBufferSize = scanner.DefaultBufferSize
+
+	// Switch to the normal split and process functions.
+	r.splitFunc = r.lineSplitFunc
+	r.processFunc = r.emitFunc
```
Comment (andrzej-stencel) [inline pkg/stanza/fileconsumer/internal/reader/reader.go:183]: This "juggling" of function references could probably be avoided altogether after the refactoring.

File: pkg/stanza/fileconsumer/internal/reader/reader.go [hunk 2]
```diff
@@ -106,9 +106,89 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 		}
 	}()
 
+	doneReadingFile := r.readHeader(ctx)
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/internal/reader/reader.go:109]: Can we just check if the function needs to be called here? Seems like calling a function and producing a log is unnecessary overhead that every user has to pay for the tiny fraction of use cases which actually have a header.

File: pkg/stanza/fileconsumer/internal/reader/reader.go [hunk 3]
```diff
@@ -106,9 +106,89 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
...
+	if doneReadingFile {
+		return
+	}
+
+	// Reset position in file to r.Offest after the header scanner might have moved it past a content token.
+	if _, err := r.file.Seek(r.Offset, 0); err != nil {
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/internal/reader/reader.go:115]: I don't think this should be in the hot path. Let's only do it once, when we detect we are done reading the header.

File: pkg/stanza/fileconsumer/internal/reader/reader.go [hunk 4]
```diff
@@ -106,9 +106,90 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
...
+	// Switch to the normal split and process functions.
+	r.splitFunc = r.lineSplitFunc
+	r.processFunc = r.emitFunc
+
+	return false
+}
+
+func (r *Reader) readContents(ctx context.Context) {
+	// Create the scanner to read the contents of the file.
+	s := scanner.New(r, r.maxLogSize, r.initialBufferSize, r.Offset, r.splitFunc)
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply pkg/stanza/fileconsumer/internal/reader/reader.go:189]: Good catch. I think it's ok to set it to the default buffer size here and not during cleanup.

The tests exist in order to demonstrate that the code is not sensitive to buffer size. It probably indicates that something should be extracted into another package, but at some point that's not the priority. If you have another way to ensure we're not dependent on a particular buffer size, we should definitely consider it.

File: pkg/stanza/fileconsumer/internal/reader/reader.go (hunk 3 above)
Comment (andrzej-stencel) [reply pkg/stanza/fileconsumer/internal/reader/reader.go]: Right, not sure what I was thinking leaving this call here. Moved to `readHeader` in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35870/commits/0f55d1cf41497186306df3e1f96e5aa98908e5d3.

File: pkg/stanza/fileconsumer/internal/reader/reader.go (hunk 2 above)
Comment (andrzej-stencel) [reply pkg/stanza/fileconsumer/internal/reader/reader.go]: It's a debug log, but I'm fine removing it. Changed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35870/commits/7f56b47a5f1dafcc01f362aefcff0366ca852837. Please let me know if this is what you had in mind.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**